wsmaster/che-core-api-factory-bitbucket-server/src/main/java/org/eclipse/che/security/oauth1/BitbucketServerOAuth1AuthorizationHeaderSupplier.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.security.oauth1;
+
+import com.google.common.base.Strings;
+import javax.inject.Inject;
+import org.eclipse.che.api.factory.server.bitbucket.server.AuthorizationHeaderSupplier;
+import org.eclipse.che.api.factory.server.scm.exception.ScmUnauthorizedException;
+import org.eclipse.che.commons.env.EnvironmentContext;
+import org.eclipse.che.commons.subject.Subject;
+
+/**
+ * Implementation of @{@link AuthorizationHeaderSupplier} that is used @{@link
+ * BitbucketServerOAuthAuthenticator} to compute authorization headers.
+ */
+public class BitbucketServerOAuth1AuthorizationHeaderSupplier
+    implements AuthorizationHeaderSupplier {
+  private final BitbucketServerOAuthAuthenticator authenticator;
+
+  @Inject
+  public BitbucketServerOAuth1AuthorizationHeaderSupplier(
+      BitbucketServerOAuthAuthenticator authenticator) {
+    this.authenticator = authenticator;
+  }
+
+  @Override
+  public String computeAuthorizationHeader(String requestMethod, String requestUrl)
+      throws ScmUnauthorizedException {
+    try {
+      Subject subject = EnvironmentContext.getCurrent().getSubject();
+      String authorizationHeader =
+          authenticator.computeAuthorizationHeader(subject.getUserId(), requestMethod, requestUrl);
+      if (Strings.isNullOrEmpty(authorizationHeader)) {
+        throw new ScmUnauthorizedException(
+            subject.getUserName()
+                + " is not authorized in "
+                + authenticator.getOAuthProvider()
+                + " OAuth1 provider");
+      }
+      return authorizationHeader;
+    } catch (OAuthAuthenticationException e) {
+      throw new ScmUnauthorizedException(e.getMessage(), e);
```

Comment:
Is this place is where we will end up if the OAuth is not configured? In that case we need another exception type here, because OAuth flow is starting when we see that user in not authorized on BB (`ScmUnauthorizedException`), and than we try OAuth, found it broken and again same exception. This will be hard to process correclty. My proposal is separate exception for Oauth misconfiruration state.

Reply:
sounds reasonable to me

Reply:
>Is this place is where we will end up if the OAuth is not configured? 

No. That would happen in `BitbucketServerOAuthAuthenticatorProvider` and `BitbucketServerApiProvider`. We will have `NoopOAuthAuthenticator` and `NoopBitbucketServerApiClient` if Bitbucket OAuth1 is not configured.  In this case, an instance of  `BitbucketServerOAuth1AuthorizationHeaderSupplier` will not be created. However, you might be right 
`ScmUnauthorizedException` might be a wrong reaction on `OAuthAuthenticationException` (As far as I can see it can happen if `java.security.GeneralSecurityException` is thrown)   what about `ScmCommunicationException ` or `UnknownScmProviderException` ?

Reply:
`ScmCommunicationException` sounds reasonable enough imo

Reply:
done